torchao/dtypes/uintx/gemlite_layout.py

```diff
@@ -0,0 +1,372 @@
+            int_data,
+            scale,
+            zero_point,
+            bias=None,
+            fma_mode=False,
+            packing_bitwidth=_layout.packing_bitwidth,
+            contiguous=_layout.contiguous,
+        )
+
+        gemlite_kwargs = {
+            "out_features": out_features,
+            "scale_activations": gemlite_linear.scale_activations,
+            "meta_args": gemlite_linear.get_meta_args(),
+        }
+
+        packed_weight, scale, zero_point = gemlite_linear.get_tensor_args()
+
+        return cls(packed_weight, scale, zero_point, gemlite_kwargs, _layout)
+
+    def to(self, *args, **kwargs):
+        kwargs = self._get_to_kwargs(*args, **kwargs)
+        device = kwargs["device"]
+        if not is_device("cuda", device):
+            raise ValueError(
+                f"GemliteAQTTensorImpl is only available for cuda device, can't convert to {device}"
+            )
+        return self.__class__(
+            self.packed_weight.to(device),
+            self.scale.to(device),
+            self.zero_point.to(device),
+            self.gemlite_kwargs,
+            self._layout,
+        )
+
+    def _apply_fn_to_data(self, fn):
+        return self.__class__(
+            fn(self.packed_weight),
+            fn(self.scale),
+            fn(self.zero_point),
+            self.gemlite_kwargs,
+            self._layout,
+        )
+
+    def get_plain(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        dq = (
+            _linear_fp_act_int4_weight_gemlite_impl(
+                torch.eye(
+                    self.scale.shape[0] * self._layout.group_size,
+                    device=self.device,
+                    dtype=self.scale.dtype,
+                ),
+                self,
+            )
+            .t()
+            .contiguous()
+        )
+
+        quant_kwargs = get_gemlite_quant_kwargs(
+            self._layout.bit_width, self._layout.group_size
+        )
+        quant_kwargs["output_dtype"] = quant_kwargs.pop("target_dtype")
+        for key in ["mapping_type", "eps", "zero_point_dtype"]:
+            del quant_kwargs[key]
+
+        int_data = quantize_affine(
+            dq,
+            scale=self.scale,
+            zero_point=self.zero_point,
+            **quant_kwargs,
+        )
+
+        return int_data, self.scale, self.zero_point
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        kwargs = {} if kwargs is None else kwargs
+
+        # we don't handle transpose operations and just ignore them. In practice the only
+        # reason a transpsoe should occur is because the functional linear
+        # op can decompose into e.g. transpose + addmm so since we want
+        # to use the gemlite matmul kernel, which expects teh weight to be passed in as is,
+        # we ignore the transpose
+        if func is aten.detach.default or func is aten.t.default:
+            return return_and_correct_aliasing(
```

Comment:
would be good to record the transposed state probably: https://github.com/pytorch/ao/blob/200589beb900d4a39a880d4e31b942bed9d31f12/torchao/dtypes/uintx/tensor_core_tiled_layout.py#L306 in case it's used later

Reply:
i think this may be better to be something handled by AQT itself rather than the tensor impl, in reality we don't want transpose causing packing and repacking of the weight, thus it should be tracked by the top level where the shape is actually being changed.